app/api/auth/create_access_token.go

```diff
@@ -1,119 +1,169 @@
 package auth
 
 import (
-	"context"
+	"errors"
 	"net/http"
-	"sync"
+	"strings"
 	"time"
 
-	"golang.org/x/oauth2"
-
 	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+	"golang.org/x/oauth2"
 
 	"github.com/France-ioi/AlgoreaBackend/app/auth"
 	"github.com/France-ioi/AlgoreaBackend/app/database"
-	"github.com/France-ioi/AlgoreaBackend/app/logging"
+	"github.com/France-ioi/AlgoreaBackend/app/domain"
+	"github.com/France-ioi/AlgoreaBackend/app/loginmodule"
 	"github.com/France-ioi/AlgoreaBackend/app/service"
 )
 
-type userIDsInProgressMap sync.Map
-
-func (m *userIDsInProgressMap) withLock(userID int64, r *http.Request, f func() error) error {
-	userMutex := make(chan bool)
-	defer close(userMutex)
-	userMutexInterface, loaded := (*sync.Map)(m).LoadOrStore(userID, userMutex)
-	// retry storing our mutex into the map
-	for ; loaded; userMutexInterface, loaded = (*sync.Map)(m).LoadOrStore(userID, userMutex) {
-		select { // like mutex.Lock(), but with cancel/deadline
-		case <-userMutexInterface.(chan bool): // it is much better than <-time.After(...)
-		case <-r.Context().Done():
-			logging.GetLogEntry(r).Warnf("The request is canceled: %s", r.Context().Err())
-			return r.Context().Err()
-		}
-	}
-	defer (*sync.Map)(m).Delete(userID)
-
-	return f()
-}
-
-var userIDsInProgress userIDsInProgressMap
-
 // swagger:operation POST /auth/token auth accessTokenCreate
 // ---
-// summary: Create a new access token
-// description: Creates a new access token (locally for temporary users or via the login module for normal users) and
-//              saves it in the DB keeping only the input token (from authorization headers) and the new token.
-//              Since the login module responds with both access and refresh tokens, the service updates the user's
-//              refresh token in this case as well.
+// summary: Create or refresh an access token
+// description:
+//     If the "Authorization" header is not given, the service converts the given code into tokens,
+//     creates or updates the authenticated user in the DB with the data returned by the login module,
+//     and saves new access & refresh tokens into the DB as well.
+//
```

Comment:
There should be a word about `code_verifier`:

> If the "Authorization" header is not given, the service converts the given OAuth2 authorization code into tokens,  creates or updates the authenticated user in the DB with the data returned by the login module,
> and saves new access & refresh tokens into the DB as well.
> If OAuth2 authentication has used the PKCE extension, the '{code_verifier}' should be provided so it can be sent together with the '{code}' to the authentication server.

Reply:
fixed